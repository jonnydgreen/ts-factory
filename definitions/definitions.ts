import { BlockInput } from './block/block.type.ts';
import { FunctionDeclarationInput } from './function-declaration/function-declaration.type.ts';
import { IdentifierInput } from './identifier/identifier.type.ts';
import { KeywordTypeNodeInput } from './keyword-type-node/keyword-type-node.type.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { SourceFileInput } from './source-file/source-file.type.ts';

// Definition
export type Definition =
  | SourceFileInput
  | StatementInput
  | ExpressionInput
  | TypeNodeInput
  | ModifierInput;

// Member Name
export type MemberNameInput = IdentifierInput;

// Statement Inputs
export type StatementInput = FunctionDeclarationInput | BlockInput;

// Expression Inputs
export type ExpressionInput = IdentifierInput;

// Type Node Inputs
export type TypeNodeInput = KeywordTypeNodeInput;

// Input
export type Input = SourceFileInput;
