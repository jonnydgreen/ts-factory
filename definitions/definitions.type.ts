import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type {
  ExpressionInput,
  MemberNameInput,
  PropertyNameInput,
  StatementInput,
  TypeElementInput,
  TypeNodeInput,
} from './definitions.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { TokenInput } from './token/token.type.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  T extends string | ts.PropertyName ? string | PropertyNameInput :
  T extends ts.Modifier ? ModifierInput :
  T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.TypeElement ? TypeElementInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  T extends ts.Token<ts.SyntaxKind> ? TokenInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;

export type PickDefinitionFields<T> = Omit<
  T,
  '__instructions' | 'leadingTrivia' | 'kind'
>;

export type DefinitionFields<T> = keyof PickDefinitionFields<T>;
