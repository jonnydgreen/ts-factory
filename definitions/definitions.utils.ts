import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type { ExpressionInput, MemberNameInput, StatementInput } from './definitions.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  // T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;
