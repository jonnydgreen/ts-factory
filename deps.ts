// Path
import { join, resolve } from 'https://deno.land/std@0.147.0/path/mod.ts';
export const Path = {
  join,
  resolve,
};

// TS
export { ts } from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
export * as TSM from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
