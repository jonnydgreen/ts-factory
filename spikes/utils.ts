export function assertNode<TNode>(node: unknown): asserts node is TNode {
}
