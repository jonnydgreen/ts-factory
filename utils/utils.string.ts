export class StringUtils {
  /**
   * Upper cases the first letter of an input string.
   */
  static upperFirst(input: string): string {
    return `${input[0]?.toUpperCase() ?? ''}${input.slice(1)}`;
  }
}
