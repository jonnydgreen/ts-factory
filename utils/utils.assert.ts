import { Definition } from '../definitions/definitions.ts';
import { ts, tsm } from '../deps.ts';
import { ItemOrArray } from '../instructions/instructions.type.ts';
import { GuardedType, Maybe } from '../types.ts';

export class AssertionError extends Error {
  override name = 'AssertionError';
  constructor(message: string) {
    super(message);
  }
}

export function assertNever(input: never): never {
  throw new AssertionError(`Input ${input} not supported`);
}

export function assertDefinitionKind(
  definition: Definition,
  kind: tsm.ts.SyntaxKind,
): asserts definition is Extract<Definition, { kind: typeof kind }> {
  if (definition.kind !== kind) {
    throw new AssertionError(
      `Definition of kind '${
        tsm.ts.SyntaxKind[definition.kind]
      }' does not match expected Node kind '${tsm.SyntaxKind[kind]}'`,
    );
  }
}

export function assertDefined<T>(input: Maybe<T>): asserts input is T {
  if (typeof input === 'undefined') {
    throw new AssertionError('Input is not defined');
  }
}

export function assertTSNodeType<
  TNode extends ts.Node,
  // deno-lint-ignore no-explicit-any
  TPredicate extends (node: any) => node is any,
>(
  node: Maybe<TNode>,
  predicate: TPredicate,
): asserts node is GuardedType<TPredicate> {
  if (typeof node === 'undefined') {
    throw new AssertionError(`Node is not defined`);
  }

  const kind = node.kind;
  if (!predicate(node)) {
    throw new AssertionError(`Invalid Node of kind ${ts.SyntaxKind[kind]}`);
  }
}

export function assertTSMNodeKind<TKind extends ts.SyntaxKind>(
  node: tsm.Node,
  kind: TKind,
): asserts node is tsm.KindToNodeMappings[TKind] {
  const nodeKind = node?.getKind();
  if (!node?.isKind?.(kind)) {
    throw new AssertionError(
      `Invalid Node of kind ${ts.SyntaxKind[nodeKind]}; expected ${ts.SyntaxKind[kind]}`,
    );
  }
}

export function isDefined<T>(
  nodeOrNodes: Maybe<ItemOrArray<T>>,
): nodeOrNodes is ItemOrArray<T> {
  if (!nodeOrNodes) {
    return false;
  }

  if (Array.isArray(nodeOrNodes) && nodeOrNodes.length === 0) {
    return false;
  }

  return true;
}

export function assertArray<T>(input: Maybe<T> | Maybe<T>[]): asserts input is T[] {
  if (!Array.isArray(input)) {
    throw new AssertionError(`Input is not of type array: ${typeof input}`);
  }
}

export function assertNotArray<T>(input: T | T[]): asserts input is T {
  if (Array.isArray(input)) {
    throw new AssertionError(`Input is of type array when it should not`);
  }
}
