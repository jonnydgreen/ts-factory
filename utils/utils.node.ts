import { ts, tsm } from '../deps.ts';
import { Maybe } from '../types.ts';

export function shouldPrintNode(node?: ts.Node): Maybe<string> {
  if (node) {
    return printNode(node);
  }
}

export function printNode(node: ts.Node): string {
  return tsm.printNode(node, { removeComments: false });
}
