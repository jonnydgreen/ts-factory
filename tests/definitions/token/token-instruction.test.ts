import { Definition } from '../../../definitions/definitions.ts';
import { TokenInputKind } from '../../../definitions/token/token.type.ts';
import { ts, tsm } from '../../../deps.ts';
import { buildNodeFromDefinition } from '../../../instructions/instructions.utils.ts';
import { assertSnapshot, blocks } from '../../../test.deps.ts';
import { createTestName, TestDefinition } from '../../test-utils.ts';

blocks.describe('Token', () => {
  const tokens: [TokenInputKind, ...string[]][] = [
    [ts.SyntaxKind.SuperKeyword],
    [ts.SyntaxKind.ThisKeyword],
    [ts.SyntaxKind.NullKeyword],
    [ts.SyntaxKind.TrueKeyword],
    [ts.SyntaxKind.FalseKeyword],
    [ts.SyntaxKind.OpenBraceToken, 'OpenBraceToken'],
    [ts.SyntaxKind.CloseBraceToken],
    [ts.SyntaxKind.OpenParenToken],
    [ts.SyntaxKind.CloseParenToken],
    [ts.SyntaxKind.OpenBracketToken],
    [ts.SyntaxKind.CloseBracketToken],
    [ts.SyntaxKind.DotToken],
    [ts.SyntaxKind.DotDotDotToken],
    [ts.SyntaxKind.SemicolonToken],
    [ts.SyntaxKind.CommaToken],
    [ts.SyntaxKind.QuestionDotToken],
    [ts.SyntaxKind.LessThanToken, 'LessThanToken'],
    [ts.SyntaxKind.LessThanSlashToken],
    [ts.SyntaxKind.GreaterThanToken],
    [ts.SyntaxKind.LessThanEqualsToken],
    [ts.SyntaxKind.GreaterThanEqualsToken],
    [ts.SyntaxKind.EqualsEqualsToken],
    [ts.SyntaxKind.ExclamationEqualsToken],
    [ts.SyntaxKind.EqualsEqualsEqualsToken],
    [ts.SyntaxKind.ExclamationEqualsEqualsToken],
    [ts.SyntaxKind.EqualsGreaterThanToken],
    [ts.SyntaxKind.PlusToken],
    [ts.SyntaxKind.MinusToken],
    [ts.SyntaxKind.AsteriskToken],
    [ts.SyntaxKind.AsteriskAsteriskToken],
    [ts.SyntaxKind.SlashToken],
    [ts.SyntaxKind.PercentToken],
    [ts.SyntaxKind.PlusPlusToken],
    [ts.SyntaxKind.MinusMinusToken],
    [ts.SyntaxKind.LessThanLessThanToken],
    [ts.SyntaxKind.GreaterThanGreaterThanToken],
    [ts.SyntaxKind.GreaterThanGreaterThanGreaterThanToken],
    [ts.SyntaxKind.AmpersandToken],
    [ts.SyntaxKind.BarToken],
    [ts.SyntaxKind.CaretToken],
    [ts.SyntaxKind.ExclamationToken],
    [ts.SyntaxKind.TildeToken],
    [ts.SyntaxKind.AmpersandAmpersandToken],
    [ts.SyntaxKind.BarBarToken],
    [ts.SyntaxKind.QuestionQuestionToken],
    [ts.SyntaxKind.QuestionToken],
    [ts.SyntaxKind.ColonToken],
    [ts.SyntaxKind.AtToken],
    [ts.SyntaxKind.BacktickToken],
    [ts.SyntaxKind.HashToken],
    [ts.SyntaxKind.EqualsToken, 'EqualsToken'],
    [ts.SyntaxKind.PlusEqualsToken, 'PlusEqualsToken'],
    [ts.SyntaxKind.MinusEqualsToken],
    [ts.SyntaxKind.AsteriskEqualsToken],
    [ts.SyntaxKind.AsteriskAsteriskEqualsToken],
    [ts.SyntaxKind.SlashEqualsToken],
    [ts.SyntaxKind.PercentEqualsToken],
    [ts.SyntaxKind.LessThanLessThanEqualsToken],
    [ts.SyntaxKind.GreaterThanGreaterThanEqualsToken],
    [ts.SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken],
    [ts.SyntaxKind.AmpersandEqualsToken],
    [ts.SyntaxKind.BarEqualsToken],
    [ts.SyntaxKind.CaretEqualsToken],
    [ts.SyntaxKind.Unknown, 'Unknown'],
    [ts.SyntaxKind.EndOfFileToken],
  ];
  const definitions = tokens.map<TestDefinition<Definition>>(
    ([token, tokenNameOverride]) => ({
      name: createTestName(
        `should process a ${tokenNameOverride ?? ts.SyntaxKind[token]} Token`,
      ),
      input: { kind: token },
    }),
  );

  for (const definition of definitions) {
    blocks.it({
      name: definition.name,
      fn: async (t) => {
        // Act
        const node = buildNodeFromDefinition(definition.input);

        // Assert
        await assertSnapshot(t, tsm.printNode(node));
      },
      ignore: definition.ignore,
      only: definition.only,
    });
  }
});
