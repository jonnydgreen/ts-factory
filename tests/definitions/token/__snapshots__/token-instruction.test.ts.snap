export const snapshot = {};

snapshot[`Token > should process a SuperKeyword Token 1`] = `"super"`;

snapshot[`Token > should process a ThisKeyword Token 1`] = `"this"`;

snapshot[`Token > should process a NullKeyword Token 1`] = `"null"`;

snapshot[`Token > should process a TrueKeyword Token 1`] = `"true"`;

snapshot[`Token > should process a FalseKeyword Token 1`] = `"false"`;

snapshot[`Token > should process a OpenBraceToken Token 1`] = `"{"`;

snapshot[`Token > should process a CloseBraceToken Token 1`] = `"}"`;

snapshot[`Token > should process a OpenParenToken Token 1`] = `"("`;

snapshot[`Token > should process a CloseParenToken Token 1`] = `")"`;

snapshot[`Token > should process a OpenBracketToken Token 1`] = `"["`;

snapshot[`Token > should process a CloseBracketToken Token 1`] = `"]"`;

snapshot[`Token > should process a DotToken Token 1`] = `"."`;

snapshot[`Token > should process a DotDotDotToken Token 1`] = `"..."`;

snapshot[`Token > should process a SemicolonToken Token 1`] = `";"`;

snapshot[`Token > should process a CommaToken Token 1`] = `","`;

snapshot[`Token > should process a QuestionDotToken Token 1`] = `"?."`;

snapshot[`Token > should process a LessThanToken Token 1`] = `"<"`;

snapshot[`Token > should process a LessThanSlashToken Token 1`] = `"</"`;

snapshot[`Token > should process a GreaterThanToken Token 1`] = `">"`;

snapshot[`Token > should process a LessThanEqualsToken Token 1`] = `"<="`;

snapshot[`Token > should process a GreaterThanEqualsToken Token 1`] = `">="`;

snapshot[`Token > should process a EqualsEqualsToken Token 1`] = `"=="`;

snapshot[`Token > should process a ExclamationEqualsToken Token 1`] = `"!="`;

snapshot[`Token > should process a EqualsEqualsEqualsToken Token 1`] = `"==="`;

snapshot[`Token > should process a ExclamationEqualsEqualsToken Token 1`] = `"!=="`;

snapshot[`Token > should process a EqualsGreaterThanToken Token 1`] = `"=>"`;

snapshot[`Token > should process a PlusToken Token 1`] = `"+"`;

snapshot[`Token > should process a MinusToken Token 1`] = `"-"`;

snapshot[`Token > should process a AsteriskToken Token 1`] = `"*"`;

snapshot[`Token > should process a AsteriskAsteriskToken Token 1`] = `"**"`;

snapshot[`Token > should process a SlashToken Token 1`] = `"/"`;

snapshot[`Token > should process a PercentToken Token 1`] = `"%"`;

snapshot[`Token > should process a PlusPlusToken Token 1`] = `"++"`;

snapshot[`Token > should process a MinusMinusToken Token 1`] = `"--"`;

snapshot[`Token > should process a LessThanLessThanToken Token 1`] = `"<<"`;

snapshot[`Token > should process a GreaterThanGreaterThanToken Token 1`] = `">>"`;

snapshot[`Token > should process a GreaterThanGreaterThanGreaterThanToken Token 1`] = `">>>"`;

snapshot[`Token > should process a AmpersandToken Token 1`] = `"&"`;

snapshot[`Token > should process a BarToken Token 1`] = `"|"`;

snapshot[`Token > should process a CaretToken Token 1`] = `"^"`;

snapshot[`Token > should process a ExclamationToken Token 1`] = `"!"`;

snapshot[`Token > should process a TildeToken Token 1`] = `"~"`;

snapshot[`Token > should process a AmpersandAmpersandToken Token 1`] = `"&&"`;

snapshot[`Token > should process a BarBarToken Token 1`] = `"||"`;

snapshot[`Token > should process a QuestionQuestionToken Token 1`] = `"??"`;

snapshot[`Token > should process a QuestionToken Token 1`] = `"?"`;

snapshot[`Token > should process a ColonToken Token 1`] = `":"`;

snapshot[`Token > should process a AtToken Token 1`] = `"@"`;

snapshot[`Token > should process a BacktickToken Token 1`] = `"\`"`;

snapshot[`Token > should process a HashToken Token 1`] = `"#"`;

snapshot[`Token > should process a EqualsToken Token 1`] = `"="`;

snapshot[`Token > should process a PlusEqualsToken Token 1`] = `"+="`;

snapshot[`Token > should process a MinusEqualsToken Token 1`] = `"-="`;

snapshot[`Token > should process a AsteriskEqualsToken Token 1`] = `"*="`;

snapshot[`Token > should process a AsteriskAsteriskEqualsToken Token 1`] = `"**="`;

snapshot[`Token > should process a SlashEqualsToken Token 1`] = `"/="`;

snapshot[`Token > should process a PercentEqualsToken Token 1`] = `"%="`;

snapshot[`Token > should process a LessThanLessThanEqualsToken Token 1`] = `"<<="`;

snapshot[`Token > should process a GreaterThanGreaterThanEqualsToken Token 1`] = `">>="`;

snapshot[`Token > should process a GreaterThanGreaterThanGreaterThanEqualsToken Token 1`] = `">>>="`;

snapshot[`Token > should process a AmpersandEqualsToken Token 1`] = `"&="`;

snapshot[`Token > should process a BarEqualsToken Token 1`] = `"|="`;

snapshot[`Token > should process a LastBinaryOperator Token 1`] = `"^="`;

snapshot[`Token > should process a Unknown Token 1`] = `""`;

snapshot[`Token > should process a EndOfFileToken Token 1`] = `""`;
